openstack_exporter/utils/cinder.py

```diff
@@ -0,0 +1,363 @@
+
+    return pool_list
+
+
+def aggregate_pools(pools):
+    """Create aggregated stats for pools.
+
+    Some pools are reported by multiple drivers (shards). This will
+    aggregate the stats for those pools.
+
+    If a pool has an aggregate_id, it is an aggregate pool, meaning
+    that the same pool is reported by multiple drivers (shards).
+
+    """
+    def calc_virtual_free(available_capacity, allocated_capacity):
+        return available_capacity - allocated_capacity
+
+    def calc_free_percent(virtual_free, total_available_capacity):
+        if total_available_capacity == 0:
+            return 0
+        return math.floor((virtual_free / total_available_capacity) * 100)
+    
+    def calc_available_capacity(total_capacity, reserved_percentage):
+        reserved = float(reserved_percentage) / 100
+        reserved_capacity = math.floor(total_capacity * reserved)
+        return total_capacity - reserved_capacity
+
+    agg_pools = {}
+    for shard in pools:
+        for pool in pools[shard]:
+            if  "aggregate_id" in pool['capabilities']:
+                caps = pool['capabilities']
+                pool_name = extract_host(pool['name'], 'pool')
+                available_cap = calc_available_capacity(
+                    caps['total_capacity_gb'],
+                    caps['reserved_percentage']
+                )
+
+                if pool_name not in agg_pools:
+                    virtual_free = calc_virtual_free(
+                        available_cap,
+                        caps['allocated_capacity_gb']
+                    )
+                    agg_pools[pool_name] = {
+                        'name': pool_name,
+                        'aggregate_id': caps['aggregate_id'],
+                        'allocated_capacity_gb': caps['allocated_capacity_gb'],
+                        'free_capacity_gb': caps['free_capacity_gb'],
+                        'total_capacity_gb': caps['total_capacity_gb'],
+                        'reserved_percentage': caps['reserved_percentage'],
+                        'max_over_subscription_ratio': caps['max_over_subscription_ratio'],
+                        'thin_provisioning_support': caps['thin_provisioning_support'],
+                        'aggregate_id': caps['aggregate_id'],
+                        'virtual_free_capacity_gb': virtual_free,
+                    }
+                    if 'netapp_fqdn' in caps['custom_attributes']:
+                        agg_pools[pool_name]['netapp_fqdn'] = caps['custom_attributes']['netapp_fqdn']
+                    else:
+                        agg_pools[pool_name]['netapp_fqdn'] = "N/A"
+
+                    virtual_free = calc_virtual_free(
+                        available_cap,
+                        agg_pools[pool_name]['allocated_capacity_gb']
+                    )
+                    agg_pools[pool_name]['free_percent'] = calc_free_percent(
+                        virtual_free, available_cap
+                    )
```

Comment:
We calculate `virtual_free` twice in this `if` - why?

One more question: I might remember wrongly, but didn't we in Cinder had to sum up the allocated capacity over the different pools belonging to the same aggregate_id? Why don't we have to do it here i.e. only add the pool once with all its data in that one `if` - did we maybe miss an `else`?

Reply:
The calculated aggregate allocated capacity is already baked in to each aggregate pool by the scheduler.  So when the call to the scheduler to fetch the stats, it's already in each aggregated pool.

I removed the 2nd call to calculate free cap.